git-remote-ipfs/src/Network/IPFS/Git/RemoteHelper.hs

```diff
@@ -170,83 +188,85 @@ processPush _ localRef remoteRef = do
             Nothing -> linkedObject root' "HEAD" "refs/heads/master"
     root'' <$ ipfs (updateRemoteUrl root'')
   where
-    go !root remoteHeadRef localHeadRef
-        | Just localHeadRef == remoteHeadRef = pure ()
-        | otherwise = do
-
-        logDebug $
-            fmt ("processPush: " % fcid % " " % fcid % " " % shown)
-                root
-                localHeadRef
-                (cidToRef @Git.SHA1 <$> remoteHeadRef)
-
+    go !root localRefCid = do
+        logDebug $ fmt ("processPush: " % fcid % " " % fcid) root localRefCid
         obj <- do
-            sha  <- liftEitherRH . first CidError $ cidToRef localHeadRef
+            sha <-
+                liftEitherRH . first CidError $
+                    cidToRef @Git.SHA1 localRefCid
             logDebug $ "sha " <> Text.pack (show sha)
-            localHeadRef' <- git $ Git.resolve (sha :: Git.Ref Git.SHA1)
-            case localHeadRef' of
-                Nothing -> throwRH $
-                    UnknownLocalRef $ Text.pack (Git.toHexString sha)
-                Just  r -> do
-                    dir <- Git.gitRepoPath <$> Git.getGit
-                    git . liftIO $ Git.looseRead dir r
+            dir <- Git.gitRepoPath <$> Git.getGit
+            git . liftIO $ Git.looseRead dir sha
 
         let raw = Git.looseMarshall obj
 
         logDebug $ "BlockPut: " <> decodeUtf8With lenientDecode (L.toStrict raw)
         blockCid <- do
             cid <- ipfs $ putBlock raw
             -- check 'res' CID matches SHA
-            when (localHeadRef /= cid) $
+            when (localRefCid /= cid) $
                 throwRH
                     . CidError
                     $ sfmt ("CID mismatch: expected `" % fcid % "`, actual `" % fcid % "`")
-                           localHeadRef
+                           localRefCid
                            cid
             pure cid
 
         logDebug $ fmt ("blockCid: " % fcid) blockCid
 
         -- if loose object > 2048k, create object + link block to it
-        objCid <-
-            if L.length raw > 2048000 then
-                linkedObject root ("objects/" <> cidToText blockCid) raw
-            else
-                pure blockCid
+        when (L.length raw > 2048000) $
+            void $ linkedObject root ("objects/" <> cidToText blockCid) raw
 
-        logDebug $ fmt ("objCid: " % fcid) objCid
         -- process links
-        traverse_ (go root remoteHeadRef) (objectLinks obj)
+        env <- ask
+        liftIO . forConcurrently_ (objectLinks obj) $
+            either (throwM . ConcurrentException) pure <=<
+                runRemoteHelper env . go root
```

Comment:
I’m a bit concerned about the potential exponential growth of the number of threads. Couldn’t we use a simple semaphore to limit the number of concurrent requests?

Reply:
The number of threads is in the 1000s, so that is not much of a concern atm imho. I realise, however, that `http-client` doesn't have an upper limit on the number of connections (I thought it had), so we might end up creating 1000s of HTTP connections. I guess a semaphore _inside_ `Concurrently` would be a compromise. That way, we still get the exception semantics from `async` for free.

Reply:
4b113d15e691d33bb79c52c698621b4f38e925bf

Reply:
Cool 😎